### ISSUE TITLE
About how to use GraphiQL and SwaggerUI in Headless Discovery Web

### ISSUE THREAD

**Issue by nhpatt** [liferay/liferay-frontend-guidelines#155]
Hi!

We want to use [GraphiQL](https://github.com/graphql/graphiql) and [SwaggerUI](https://github.com/swagger-api/swagger-ui/tree/master/flavors/swagger-ui-react) for a react portlet that is not available (hidden) in liferay portal but is served from a REST API instead (/o/api).

Both libraries are huge and have integration problems with the current bundler, the first one uses ESNext, second one has multiple problems (lodash, node-fetch).

We can avoid integrating with Swagger (and use our own React app) but we do want to use GraphiQL... how would be the best way of doing it?

I see several options:

- Use the new bundler
- Have our own custom webpack configuration, and serve the html in the API directly. We don't need to be present in the portal, we can serve a "static" html.
- Change to a servlet and serve a static html (we even don't have to be a portlet, it can be useful for a documentation site but is not required) referencing liferay react copy somehow.
- 🏃 

What do you think?

**Comment by wincent** [comment]
Some initial thoughts.

A few years back when I was working on GraphiQL, the idea for it was that it be a batteries-included, self-contained "IDE" kind of thing that could be easily integrated by dropping a bundle somewhere. That's why the [demo links in the REAME](https://github.com/graphql/graphiql/tree/master/packages/graphiql#demos) point at pre-bundled versions and there's info there about [how to run from CDN](https://github.com/graphql/graphiql/tree/master/packages/graphiql#cdn-bundle).

Even express-graphql, the official "reference" GraphQL server implementation in JS, embeds GraphiQL [via prebuilt bundles](https://github.com/graphql/express-graphql/blob/d634e419663fb8e3133fe5fada47809192e6e5e2/src/renderGraphiQL.js).

Fun fact: internally at Facebook, the schema was so large (many megabytes) that it dwarfed the size of GraphiQL and all its dependencies, even though the dependency footprint is fat. 😂 IIRC, the internal GraphiQL instance at FB was a pre-built bundle with some tweaks on top that didn't go through the normal packager that public-facing JS would go through.

All of which is to say, I would probably just make a pre-built bundle for this kind of thing. @izaera may still want to use this as a stress-test though for v3 of the Bundler. Not sure.

**Comment by izaera** [comment]
I would say the only way to go (if we want to use the bundler) is to try v3. This means I would totally discard bundler v2.

Maybe we can do a spike with bundler v3 and see what happens. Then, if it fails, depending on how long it takes to fix it, we may fallback to a custom webpack build.

**Comment by nhpatt** [comment]
A pre-built bundle sounds good for me :) About swagger, should we follow the same solution?

How would be serve the pre built bundle, can we include minimified JS in a module or we should built it with a custom webpack configuration?

**Comment by jbalsas** [comment]
Seeing as this is mostly an external app... I would go full webpack bundles on this. I would really go for options 2 or 3

> - Have our own custom webpack configuration, and serve the html in the API directly. We don't need to be present in the portal, we can serve a "static" html.
> - Change to a servlet and serve a static html (we even don't have to be a portlet, it can be useful for a documentation site but is not required) referencing liferay react copy somehow.

Just built an app any way you want and serve it. If being a portlet or running inside portal is not a requirement, it will probable save us a lot of headaches.

One secret project we have (I don't think @wincent knows about this yet because I haven't created the epic for it) is to provide a "Remote App" Widget. Basically, an iframe that you can configure to point to a URL with some added facilities (message bus) to integrate external apps into DXP. This is basically oriented towards the SaaS usecase, but it should also cover the "_it can be useful for a documentation site_" part for you.

**Comment by nhpatt** [comment]
Whatever you choose :)

Two important things I forgot:

- Brian does not want to serve external resources (so cdn or similar are not in the picture, we should serve ourselves the JS file)

- It should work in 7.1, 7.2... but this is just to add SwaggerUI in master/7.3 because is not needed in 7.1/7.2 so currently this is fullfilled (but we could have one type of solution in 7.3 and another in 7.1/7.2, just to take it into account).

**Comment by jbalsas** [comment]
> Brian does not want to serve external resources (so cdn or similar are not in the picture, we should serve ourselves the JS file)

Yeah, should work locally and not over the network... and some sandboxed enterprises often block external access, specially if they have private APIs or sensitive data :)

> It should work in 7.1, 7.2...

Yeah... standalone APP sounds like the bast path forward here :)

**Comment by nhpatt** [comment]
Ok, so if I understand it correctly the best path would be having a standalone aplication with her own webpack.config ? (as analytics-client-js has but to build a library) and should I reference react portal version somehow? (0 knowledge of webpack here).

**Comment by jbalsas** [comment]
No portal 😂

**Comment by nhpatt** [comment]
I've tried it here: https://github.com/nhpatt/liferay-portal/commit/506636a6a0f533412b423c0ebac2c47307a50984 It works but it has devDependencies + many many changes to yarn.lock. 

Can be excluded from the build but run with the build (yeah, I know that I'm very precise) somehow or is it more of a Peter Shin doubt?

**Comment by wincent** [comment]
Just linking here (for Liferay folks) to [some discussion](https://liferay.slack.com/archives/CNBG06JS3/p1589308474067700) we had in Slack.

It was a long thread, but to summarize some of the key points:

- You can avoid polluting the yarn.lock but that would undermine its value as a centralized registry of the dependencies that we actually rely on in liferay-portal; it doesn't matter that the GraphiQL/Swagger apps are delivered by a different mechanism from the other stuff in liferay-portal — the dependencies are still real and hiding them from the lockfile would be "sweeping them under the carpet".
- The right way to add `devDependencies` is to bring them in via `liferay-npm-scripts`. Some detailed about how this is, in theory, enforced [here](https://github.com/liferay/liferay-npm-tools/issues/453).  This does raise the barrier to adding new dependencies, but that added friction is intentional — in this case, I don't at first glance see any problem in adding the specific `devDependencies` you've added here:
    - They look to be relatively "standard" webpack dependencies.
    - We already officially support webpack builds in liferay-portal (that's why it's a feature of `liferay-npm-scripts`), so it is reasonable to extend the set of bundled related dependencies when we discover a gap that needs to be filled.
- I don't have any specific thoughts/ comments on CSS libraries; if you want to bring up specific alternatives @nhpatt we can analyze them.

But like I said, the thread was long... are there any things that I missed? @izaera @nhpatt?

**Comment by izaera** [comment]
> But like I said, the thread was long... are there any things that I
> missed? @izaera @nhpatt?
> 

I think it is a good summary :+1:

**Comment by nhpatt** [comment]
Yes, I agree on all counts. So next steps would be requesting to add those devDependencies and check if the dependencies added in the yarn.lock are fine, right?

About the css libraries, currently the devDependencies are:

```
"html-webpack-plugin": "4.3.0",
"script-ext-html-webpack-plugin": "2.1.1",
"html-inline-css-webpack-plugin": "1.8.0",
"mini-css-extract-plugin": "0.9.0"
```
I think _style-ext-html-webpack-plugin_ (the companion css library for script-ext-html-webpack-plugin), instead of _html-inline-css-webpack-plugin_ would also work fine, that I was applying the plugins in the wrong order (but I have to check).

**Comment by wincent** [comment]
Once you've got your finalized list of plug-ins, let us know and we'll publish an update to liferay-npm-scripts and we can sanity check the yarn.lock when we do the update in liferay-portal (and remove your local `devDependencies` and the suppression).
